keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,571 @@
+  // This is a REQUIRED field.
+  string volume_path = 2;
+
+  // The path where the volume is staged, if the plugin has the
+  // STAGE_UNSTAGE_VOLUME capability, otherwise empty.
+  // If not empty, it MUST be an absolute path in the root
+  // filesystem of the process serving this request.
+  // This field is OPTIONAL.
+  string staging_target_path = 3;
+}
+```
+
+```
+message NodeGetVolumeStatsResponse {
+  // This field is OPTIONAL.
+  repeated VolumeUsage usage = 1;
+  // This field is OPTIONAL.
+  repeated VolumeHealth volume_health = 1;
+}
+```
+
+```
+message VolumeUsage {
+  enum Unit {
+    UNKNOWN = 0;
+    BYTES = 1;
+    INODES = 2;
+  }
+  // The available capacity in specified Unit. This field is OPTIONAL.
+  // The value of this field MUST NOT be negative.
+  int64 available = 1;
+
+  // The total capacity in specified Unit. This field is REQUIRED.
+  // The value of this field MUST NOT be negative.
+  int64 total = 2;
+
+  // The used capacity in specified Unit. This field is OPTIONAL.
+  // The value of this field MUST NOT be negative.
+  int64 used = 3;
+
+  // Units by which values are measured. This field is REQUIRED.
+  Unit unit = 4;
+}
+```
+
+```
+message VolumeHealth {
+  // The error code describing the health condition of the volume.
+  // This field is REQUIRED.
+  string error_code = 1;
+
+  // The error message associated with the above error_code. This field is OPTIONAL.
+  string message = 2;
+}
+```
+
+Add a `GET_VOLUME_STATS_HEALTH` node capability. If driver supports this capability, it MUST fetch `volume_health` information in `NodeGetVolumeStats`.
+
+```
+message NodeGetCapabilitiesRequest {
+  // Intentionally empty.
+}
+```
+
+```
+message NodeGetCapabilitiesResponse {
+  // All the capabilities that the node service supports. This field
+  // is OPTIONAL.
+  repeated NodeServiceCapability capabilities = 1;
+}
+```
+
+```
+// Specifies a capability of the node service.
+message NodeServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+      STAGE_UNSTAGE_VOLUME = 1;
+      // If Plugin implements GET_VOLUME_STATS capability
+      // then it MUST implement NodeGetVolumeStats RPC
+      // call for fetching volume statistics.
+      GET_VOLUME_STATS = 2;
+      // See VolumeExpansion for details.
+      EXPAND_VOLUME = 3;
+      // If Plugin implements GET_VOLUME_STATS_HEALTH capability
+      // then it MUST implement NodeGetVolumeStats RPC
+      // call for fetching volume health information.
+      GET_VOLUME_STATS_HEALTH = 4;
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+```
+
+### External controller
+
+#### CSI interface
+Call GetVolume() RPC for volumes periodically to check the health condition of volumes themselves. The frequency of the check should be tunalbe. A configure option will be available in the external controller to adjust this value.
+
+#### Node down event
+* Watch node down events.
```

Comment:
Would this be watching Node status?

Reply:
Yes, it will watch node status.  In addition I think we also need to ping the node as it could be an unplanned shut down.